Allow flux_future_continue(3) to work with flux_kvs_lookup(3)

grondo opened the pull request:
This PR fixes #2127 by adding a new function `flux_future_fulfill_with(3)`, which allows a fulfilled future to be used to fulfill another future. This is what `fulfill_next()` in `composite_future.c` tried to do, but its scope was limited because it didn't have access to `flux_future_t` internals.

`flux_future_fulfill_with(3)` copies any fatal error, error result, or normal result into the destination future. It also "embeds" the source future into the target future, so that calls to `flux_future_aux_get()` can access the source future aux list. This allows futures that are a result of calls which use aux data as part of custom `get()` methods to work after `flux_future_fulfill_with(3)`.  This function steals a reference to the source future, so `flux_future_destroy()` should not be called on the source future after `flux_future_fulfill_with()` is called. (If this is confusing, the function could just take a reference instead)

The destination future also "steals" the valid result by copying the result's destructor into its result structure, nullifying the destructor in the source future. This is so that multiple fulfillment can work properly, since once `flux_future_reset()` is called on the source future, a result pending in the destination future may point to invalid memory.

Finally, `composite_future.c:fulfill_next()` was updated to use `flux_future_fulfill_with()` to get these benefits, allowing `flux_future_and_then(3)` to be used with `flux_kvs_lookup(3)` and other functions that return futures and make use of the aux data in `flux_future_t`.

garlick commented:
Nicely done.  I'm fine with pulling this in if you're ready.

trws commented:
This certainly seems to work.  The example I was playing with, re-factored to correctly represent `and_then` and correctly type the intermediate futures, now runs correctly to completion it seems:

```rust
    let mut composite = h.kvs_lookup("", 0, "thing")?
        .and_then(|fi| {
            eprintln!("kvs result:{:?}", fi.lookup_get()?);
            h.kvs_lookup("", 0, "other_thing")
        })?
        .and_then(|fi| {
            eprintln!("kvs result2:{:?}", fi.lookup_get()?);
            h.kvs_lookup("", 0, "other_thing")
        })?
        .then(|f| {
            // f is strongly typed with kvs future methods
            println!("kvs result3:{:?}", f.lookup_get().unwrap());
        })?;
```

The only odd thing, which may have nothing to do with this, is that tacking a wait_for(-1)  on the end gives me a "no such file or directory" error as soon as the composite future gets there.  If I swap that for a reactor_run, everything works fine seemingly.  That happens with or without this patch though, so I'm not sure it should delay merging this.

grondo commented:
Hold off, I'll throw in a fix for #2135 as well in here.

And I'll try to see if I can reproduce and possibly fix the issue described by @trws above as well.

grondo commented:
Forced a push here changing the ownership semantics of `flux_future_continue_with(3)`. It was insanity to have this function steal the reference to the source future `p`, especially when the function could be called from as a result of  another operation.

The function now takes a reference to the source future, and thus `fulfill_next()` in the composite future code is safe to destroy the future (releasing the implicit reference taken by `flux_future_continue(3)`, which makes more logical sense)

I'm still working on support for #2135

grondo commented:
There is actually a tricky memory management case I ran across here while trying to implement a fix for #2135.

If a user registers a `flux_future_and_then(3)` callback on a future `f`, there is *already* a fallback case when `f` is fulfilled with an error. When there is no `or_then()` callback registered, the fall through immediately uses `f` to fulfill the `next` future in the chain so that the error is propagated to either the final future in the chain or the next `or_then()` callback.

The problem is that futures are typically destroyed in their continuations immediately after the result is used, or for chained futures after `flux_future_continue (f, next)` is called. The proposed fall back (and the existing one) don't give the user a chance to destroy the future `f` after it is propagated.

I think the current code only works because the future gets accidentally destroyed after the error is propagated. We could keep this behavior,  but then there is a problem if a user decides to destroy the future *outside* of the `and_then()` callback -- in normal conditions this will work, but for the error case, there would be a double free.

Probably the only sane solution here is to make it so that a call to`flux_future_and_then(3)` or `flux_future_or_then(3)` steals the reference to the target future. This would mean that for `and_then` or `or_then` callbacks, `flux_future_destroy (f)` would not be used. If the future is fulfilled, then the implementation would destroy the future in its internal continuation. Unfulfilled futures would be destroyed by the destructor of the successor future `next`, which was returned by `flux_future_and_then` or `or_then`.

grondo commented:
> Probably the only sane solution here is to make it so that a call toflux_future_and_then(3) or flux_future_or_then(3) steals the reference to the target future.

An alternative might be to document that the default continuation for a chained future which has not had `flux_future_continue(3)` called is:

```C
flux_future_fulfill_with (next, f);
flux_future_destroy (f);
```

This avoids changing a lot of code, and also avoids holding onto the memory for chained futures longer than necessary (if we take the approach above of having the `next` future "own" the memory for the future argument to `and_then/or_then`, then memory for futures in a chain isn't released until the final future is destroyed.)

This also handles #2135 since the above propagation is now the default, overridden only by a call to `flux_future_continue(3)`.

trws commented:
I’m not sure if this will be useful information or not, but the usual pattern for a “then” type call is that the future returned by then owns the future that’s passed in, as you suggested. Rather than retaining that ownership until destruction though it’s usually destroyed by the system after the callback completes, since it has to have been fulfilled and either read or had its refcount increased in the callback.

Sent from my iPhone

On Apr 19, 2019, at 4:48 PM, Mark Grondona <notifications@github.com<mailto:notifications@github.com>> wrote:


Probably the only sane solution here is to make it so that a call toflux_future_and_then(3) or flux_future_or_then(3) steals the reference to the target future.

An alternative might be to document that the default continuation for a chained future which has not had flux_future_continue(3) called is:

flux_future_fulfill_with (next, f);
flux_future_destroy (f);

This avoids changing a lot of code, and also avoids holding onto the memory for chained futures longer than necessary (if we take the approach above of having the next future "own" the memory for the future argument to and_then/or_then, then memory for futures in a chain isn't released until the final future is destroyed.)

This also handles #2135<https://github.com/flux-framework/flux-core/issues/2135> since the above propagation is now the default, overridden only by a call to flux_future_continue(3).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/flux-framework/flux-core/pull/2130#issuecomment-485039536>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAFBFNOA4RGPQ3X6277SUZ3PRJK3PANCNFSM4HGWUQPQ>.

grondo commented:
> I’m not sure if this will be useful information or not, but the usual pattern for a “then” type call is that the future returned by then owns the future that’s passed in, as you suggested.

Thanks @trws! This is actually really helpful. Following existing convention seems like the right way to go! I was still struggling a bit with which was the right way to go.

garlick commented:
Should we consider (separate issue) changing `flux_future_then()` so that the future is implicitly destroyed when the continuation returns?  Since there's an incref, it's still user's choice to keep the future around if that's useful, as it sometimes is when the contained result is to be used elsewhere.

grondo commented:
> Should we consider (separate issue) changing flux_future_then() so that the future is implicitly destroyed when the continuation returns? Since there's an incref, it's still user's choice to keep the future around if that's useful, as it sometimes is when the contained result is to be used elsewhere.

I was thinking the same thing. Once you call `flux_future_then()` the callback owns the memory for the future. My only concern is a user has to know if the callback was actually called to know if the future was destroyed or not. However, now that I write that down, that's no different than the current situation.

trws commented:
That's exactly it, the runtime effectively "owns" the future at that point, or at least takes a reference to it and should incref it so the user can't pull it out from under you.  We effectively have only "shared" futures, in that they're all potentially reference counted, the behavior for a non-refcounted future is usually that it is moved into the next one and the caller can no longer do anything with it, it's effectively destroyed as far as their reference is concerned.  Part of this though is that "then" normally itself returns a future which can be waited on to wait for the completion of the callback or an error, which takes care of the issue you mentioned @grondo.  The normal chained workflow looks, pseudocode-wise, something like this for ownership:

```c++
fut f1 = something_async();
fut f2 = fut_then(f1, [](fut f){ // f2 now owns f1
  int res = something_with_result(fut_get(f));
  // fut_incref(f); // if we want to keep it live for some reason
  fut_fulfill(f, res);
   });
if (fut_wait(f2) < 0) // f1 dies here or before, unless incref called
  handle_error()
int res = fut_get(f2); // often future is actually destroyed by this, optional though
fut_destroy(f2);
```
The thing where our model differs right now is that it doesn't represent a value coming out of the callback unless you use and/or_then, and rather than returning a future if you want to pass one on you call continue or continue_error to do it.  It's trickier to deal with that in C unfortunately since you can't do strong typing on it, but we could probably figure something out.

grondo commented:
If we've decided to change the ownership model of base `flux_future_t` type as @trws suggests above, I'll open a separate issue on that and punt on solving #2135 here, since doing it the easy way could result in impossible-to-solve allocation issues in some cases, and I tried doing the "right" way and it quickly turned into a huge mess.

Once the base `flux_future_t` type ownership model changes, I think it will make it a lot easier to make the change suggested in #2135 without adding confusing ownership differences between then(), and_then() and or_then().

grondo commented:
I've force-pushed an branch rebased on current master and constrained the fixes to #2127

grondo commented:
One more note -- instead of fixing #2135, it would be trivial to temporarily add an assertion in the internal chained future callback that would trigger if a callback did not call either flux_future_continue() or flux_future_continue_error().

trws commented:
That seems like it would be a good stopgap.

On Apr 25, 2019, at 8:36 AM, Mark Grondona <notifications@github.com<mailto:notifications@github.com>> wrote:


One more note -- instead of fixing #2135<https://github.com/flux-framework/flux-core/issues/2135>, it would be trivial to temporarily add an assertion in the internal chained future callback that would trigger if a callback did not call either flux_future_continue() or flux_future_continue_error().

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/flux-framework/flux-core/pull/2130#issuecomment-486725172>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAFBFNKXHI2FD6H2GO3267DPSHFYTANCNFSM4HGWUQPQ>.

grondo commented:
Ok, maybe I'll attempt that here over I'm back at a computer

grondo commented:
> That seems like it would be a good stopgap.

Actually, I've followed the original plan here and have pushed a commit that auto-fulfills the next future in the chain with the current future when either `flux_future_continue()` wasn't called from the callback, or a callback wasn't called at all (i.e. if a future is fulfilled with success and no `and_then` callback exists, or fulfilled with error and no `or_then` callback exists.)

If any callback is used, then it is expected that the user will destroy the future within the callback or elsewhere. If no callback was called the future is destroyed. This not only preserves current behavior, but also will be easier to update and reason about if we change the ownership model of the base `flux_future_t` type in the future.

garlick commented:
Just trying to get up to speed on this one again.  It sounds like you covered all the cases I was concerned about (multiple fulfillment, aux collision with multiple futures of the same type in a chain).  I'll try a couple of random tests when I get to work, but looks like a great job so far.

And thanks for the inline documentation.  I really struggle with this future code and having that will help.

garlick commented:
Following up on our face to face discussion:   I was hoping to implement the single-job state synchronization (#1665) on top of `flux_job_event_watch()`, which returns a future that is fulfilled for each job event.  Ideally `flux_job_state_watch()` would internally handle events, and return a future that is fulfilled only when a watched state is reached.  It would operate as a sort of filter on the event watch fulfillments.

Could I use `flux_future_and_then()` to implement this with a composite future?  As discussed, the new `continued` flag added in this PR presents a difficulty since when an event fulfilling the event future doesn't advance the current state to one of the watched states, there should not be a fulfillment of the composite future.  With the flag, the composite future would be auto-fulfilled in that case.  Could we add a function that could be called from the event continuation to defeat that behavior for one iteration?

grondo commented:
>  Could we add a function that could be called from the event continuation to defeat that behavior for one iteration?

Yes. Perhaps this should be added regardless of whether it satisfies your current use case or not.

The easiest way to support this would be to allow a `NULL` second argument to `flux_future_continue(3)`. The current prototype for that function is:
```C
/* Set the next future for the chained future `prev` to `f`.
 *  This function steals a reference to `f` and thus flux_future_destroy()
 *  should not be called on `f`. (However, flux_future_destroy() should
 *  still be called on `prev`)
 */
int flux_future_continue (flux_future_t *prev, flux_future_t *f);
```

Already we document  (or should document if I haven't done that yet here) that if `flux_future_continue(3)` is not called on `prev` then `next` is fulfilled using the result from `prev` by default. We could document that if `prev` is "continued" with a `NULL` future, then this automatic fulfill is disabled, and the next future remains unfulfilled.

garlick commented:
That will actually work great for this case.  `flux_future_continue (NULL)` will fulfill the "state" future with the event that transitioned to that watched-for state, and the event continuation can accumulate events in its aux hash.  So upon fulfillment, not only the state but the event that led to it, and metadata encoded in the the eventlog thus far can be made evailable via a `get` function.  In particular, upon entering CLEANUP or INACTIVE, one could retrieve either `exception` or `finish`  metadata.

grondo commented:
> That will actually work great for this case. flux_future_continue (NULL) will fulfill the "state" future with the event that transitioned to that watched-for state, and the event continuation can accumulate events in its aux hash. 

Just to be clear, I was thinking
```C
/* Do *not* fulfill the "next" future in the chain on this iteration, and disable
 *  default auto-continue for chained futures that don't have flux_future_continue(3)
 *  called on them from this callback
  */
flux_future_continue (f, NULL);

/* Do fulfill the next future in the chain with the current result stored in `f`:
 */
flux_future_continue (f, f);
```

I know, that second one is weird. You can also get the same result if you just don't call `flux_future_contine()` obviously. Maybe we need specialized functions to more clear, like:

```C
#define flux_future_nocontinue (x) flux_future_continue (x, NULL)
#define flux_future_continue_next (x) flux_future_continue (x, x)
```

It seemed like perhaps you had a better idea above, though. I'm really open to anything that lessens confusion.

garlick commented:
Apologies I had it backwards.  I call nothing to fulfill the state future with with the event future (or `flux_future_continue (f, f)`, and I call `flux_future_continue (f, NULL)` to leave the state future unfulfilled.

Sorry about that.

grondo commented:
Ok, I've implemented support for the above (still have to write tests and docs), but thought you could try it out before I get any further?

(Sorry, I also rebased on current master without thinking)

garlick commented:
I was able to use the above strategy to implement a prototype state monitoring API call!  Nice, thank you!

grondo commented:
Ok, I also pushed some sanity tests into `test_composite_future.t` for `flux_future_continue (f, NULL)` and `flux_future_continue (f, f)` as well as multiple-fulfillment for chained futures.

If this approach is good I will also update documentation.

garlick commented:
Thanks for adding the tests!  LGTM!

grondo commented:
Just rebased on master and reworded `flux_future_and_then(3)` manpage to note that fulfilling the next future with the current future after a chained continuation is now the default, and document the use of `flux_future_continue (f, NULL)`.

The manpage could probably use some examples, though they might be a bit involved. Have to think about it a bit.

O/w, if this looks good I can squash down some incremental development.

garlick commented:
Looks good to me.  Please squash if you're ready.

grondo commented:
Ok, squashed.

grondo commented:
One builder failed with
```
ERROR: t1008-kvs-eventlog.t - exited with status 137 (terminated by signal 9?)
```

Looks like it just overran the grace timeout:
```
PASS: t1008-kvs-eventlog.t 6 - flux kvs eventlog append fails with invalid context
# passed all 6 test(s)
flux-start: 0 (pid 11685) Killed
1..6
```

Restarted builder.

garlick commented:
Thanks!

grondo commented:
Thanks!
